Changed the error message displayed when a select element has props.multiple set to true and value set to null

Hendeca opened the pull request:
Addresses issue #9038 

I changed the error message shown when a select element with props.multiple set to true has a value of `null` to recommend an empty array instead of an empty string. Given that "the empty string" is a common term but "the empty array" is less so, I also changed the null input value error message shown for textarea and input to use "an empty string" instead of "the empty string" and changed tests to match.

gaearon commented on src/renderers/dom/shared/hooks/ReactDOMNullInputValuePropHook.js:
```diff
@@ -29,9 +29,10 @@ function validateProperties(type, props) {
+        'Consider using an empty %s to clear the component or `undefined` ' +
         'for uncontrolled components.%s',
       type,
+      type === 'select' ? 'array' : 'string',
```
This doesn't seem to check for .multiple.

gaearon commented:
Can you please rebase the branch of master instead of merging? This way unrelated commits won't be in the history.

Hendeca commented:
@gaearon Okay I've squashed my commits and rebased so you should now see only my changes.  I also added the check for the property `multiple` and added a test for that case.

Hendeca commented:
I just rebased with master to fix the merge conflicts that arose. @gaearon Is there anything else I need to do for this pull request?

gaearon commented on packages/react-dom/src/__tests__/ReactDOMSelect-test.js:
```diff
@@ -524,6 +524,25 @@ describe('ReactDOMSelect', () => {
+
+    expectDev(console.error.calls.argsFor(0)[0]).toContain(
+      '`value` prop on `select` should not be null. ' +
+        'Consider using an empty array when multiple is ' +
```
Nit: value and select are backticked but multiple isn't. Let's backtick it too.

gaearon commented on packages/react-dom/src/__tests__/ReactDOMSelect-test.js:
```diff
@@ -524,6 +524,25 @@ describe('ReactDOMSelect', () => {
+    expectDev(console.error.calls.argsFor(0)[0]).toContain(
+      '`value` prop on `select` should not be null. ' +
+        'Consider using an empty array when multiple is ' +
+        'set to true to clear the component or `undefined` ' +
```
Same for true.

Hendeca commented:
@gaearon I've added the requested changes. Let me know if you need anything else, thanks!

gaearon commented on packages/react-dom/src/shared/ReactDOMNullInputValuePropHook.js:
```diff
@@ -23,17 +26,25 @@ function validateProperties(type, props) {
-  if (props != null && props.value === null && !didWarnValueNull) {
+
+  var isMultipleSelect = type === 'select' && props.multiple === true;
+  var errorType = isMultipleSelect ? 'multiple' : 'default';
```
I don't think this is important. I'd prefer a single boolean to deduplicate (even though deduplicating one will influence the other). They're rare enough.

Then we would move the `isMultipleSelect` check inside the condition. So it's only checked in error case.

Hendeca commented:
@gaearon Thanks, sorry I wasn't able to get to this faster!
